### PR TITLE
Fix excessive requests for iModel thumbnails

### PR DESCRIPTION
**Pull request by roluk**
We now cache responses from iTwin API where the returned resource is assumed to never change (currently only applies to iModel thumbnails).

We also deduplicate active requests for the same resource in two second windows.

The component that highlighted the lack of caching has been optimised to render fewer times.

### CODE REVIEW

File: app/frontend/src/app/ITwinApi.ts [hunk 1]
```diff
@@ -86,82 +69,119 @@ interface GeoCoordinates {
...
-    return;
-  }
-
-  return response.blob();
+export function getIModelThumbnail(iModelId: string): CallITwinApiArgs<Blob> {
+  return {
+    endpoint: `imodels/${iModelId}/thumbnail?size=small`,
+    immutable: true,
+    postProcess: async (response) => response.blob(),
+  };
```
Comment (grigasp) [inline app/frontend/src/app/ITwinApi.ts:109]: should names of these functions be changed to something like `createIModelThumbnailRequestArgs`?

Comment (roluk) [reply app/frontend/src/app/ITwinApi.ts:109]: I understand what you're getting at, these functions don't perform the fetch action themselves even though their names imply otherwise.

Despite this, I think having short and concise names for these functions outweighs a possible one-time confusion. The fact that they are not `async` makes it so programmers will receive a lint error when they type `await getIModel(iModelId);`

Comment (grigasp) [reply app/frontend/src/app/ITwinApi.ts:109]: I disagree - functions should always tell what they do, hopefully, but not necessarily with a short name. Auto-complete tools are here to help developers avoid having to type long names.

Comment (roluk) [reply app/frontend/src/app/ITwinApi.ts:109]: The problem I am trying to solve is related to reading the code, not writing it, because we have already solved the latter with contemporary IDEs, as you've already mentioned.

The pattern of `something_theonlyrelevantpartinthemiddle_something` is somewhat hard to parse quickly at a glance. We can try to come up with a different formulation that adds minimal cruft only on one side, but first I need to give you the context for why the names currently are like this.

Each REST resource has a set of request methods: GET, POST, UPDATE, DELETE. The simplest possible way to give a name to a request is to combine a method and a resource name: `getIModel`, `postIModelThumbnail` and so on.

Ideally the call site would look like

```ts
await callITwinApi("getIModel", authClient);
```

However, our requests usually take in parameters like iModel id, therefore we need to somehow pass these parameter in.

My goal was to make the API and the implementation as simple as possible, so I chose to give each REST action a constructor in form of `methodResource(args)`. The name of constructor could start with a capital letter like [all new built-in web APIs do](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Symbol#description), but I thought that would look kind of weird to you as you're not well versed in JavaScript ecosystem and the different ways people express popular APIs.

Would 

```ts
await callITwinApi(GetIModel(params), authClient);
```

work better for you?

Comment (grigasp) [reply app/frontend/src/app/ITwinApi.ts:109]: No.

I agree with everything you wrote regarding request methods and the ideal approach. I disagree with the chosen way to pass method + args to the `callITwinApi` function.

Here are the rules I'd like us to keep:
1. Functions should start with a verb: "get", "post", "create", etc.
2. Functions should tell what they do.
3. Classes should be nouns: "Handler", "Factory", "Symbol", etc.

You have functions that create input arguments for `callITwinApi` which break rule #2. 

You're suggesting us to treat these functions as a constructor by starting them with an uppercase letter, but then it would break rule #3. Note that the `Symbol` example you gave doesn't break this rule. This suggestion, combined with the `Request` suffix, which you suggested in our private chat, could work for me.

BTW, the reason this discussion was raised in the first place was that you changed from `getSomething()` calling `callITwinApi(...)` to `callITwinApi(...)` accepting result of `getSomething()` as its argument. What was the purpose of this change?

Comment (roluk) [reply app/frontend/src/app/ITwinApi.ts:109]: > [...] What was the purpose of this change?

The initial idea was to separate concerns of fetching a resource and caching the response. The end result manifested in simpler code because error and authentication handling became part of centralised `callITwinApi` call.

We could go back to a single function call (`getIModel({ iModelId, authorizationClient })`), but that would add more code than necessary to accomplish the task at hand.

Comment (grigasp) [reply app/frontend/src/app/ITwinApi.ts:109]: > but that would add more code than necessary to accomplish the task at hand

How?

Currently the call looks like this:
```ts
const response = await callITwinApi(getUserProjects("representation", searchQuery), userAuthorizationClient);
```

Calling `callITwinApi` from `getUserProjects` would change the above to:
```ts
const response = await getUserProjects("representation", searchQuery, userAuthorizationClient);
```
This doesn't look more complex.

From the implementation side of things, the `callITwinApi` doesn't need to change at all, except for maybe removing the `export`. The `getUserProjects` function would simply call `callITwinApi` with the object it's now creating. Doesn't look more complex either.

Comment (roluk) [reply app/frontend/src/app/ITwinApi.ts]: Sorry, I was thinking what would happen when `callITwinApi` gains more options to control the request (e.g. retry amount, timeouts), but of course this can be supported pretty easily by separating these shared parameters like `getIModel(iModelId, sharedArgs)`. That's what I've done in 9f6d6512e043a568c2a7565c92b4c10f836249de.

**Review by grigasp — APPROVED** [review]
